File: LAB/Gramola_Backend/src/main/java/edu/uclm/esi/Gramola_Backend/services/AppUrlService.java
/* CHECK
 * Servicio encargado de gestionar las URLs configurables de la aplicación.
 *
 * Esta clase lee las URLs desde la tabla app_url usando AppUrlDao.
 * La uso para no tener URLs hardcodeadas en el código, tanto en backend
 * como en frontend.
 *
 * También tiene un método auxiliar para unir URLs sin crear barras dobles.
 */
package edu.uclm.esi.Gramola_Backend.services;

import java.util.Map;
import java.util.stream.Collectors;

import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import edu.uclm.esi.Gramola_Backend.dao.AppUrlDao;
import edu.uclm.esi.Gramola_Backend.model.AppUrl;

@Service
public class AppUrlService {

    /*
     * DAO que me permite consultar la tabla app_url.
     */
    private final AppUrlDao appUrlDao;

    /*
     * Inyecto AppUrlDao por constructor.
     * Así Spring me proporciona automáticamente el repositorio.
     */
    public AppUrlService(AppUrlDao appUrlDao) {
        this.appUrlDao = appUrlDao;
    }

    /*
     * Leo una URL desde base de datos usando su clave.
     * Si no existe o está desactivada, lanzo un error para detectarlo rápido.
     */
    public String getUrl(String id) {
        AppUrl appUrl = appUrlDao.findByIdAndActiveTrue(id)
            .orElseThrow(() ->
                new ResponseStatusException(
                    HttpStatus.INTERNAL_SERVER_ERROR,
                    "No existe la URL activa en base de datos: " + id
                )
            );

        return appUrl.getUrlValue();
    }

    /*
     * Devuelvo todas las URLs activas en formato clave-valor.
     * Esto lo usa Angular para cargar las URLs desde el backend.
     */
    public Map<String, String> getAllActiveUrls() {
        return appUrlDao.findByActiveTrue()
            .stream()
            .collect(Collectors.toMap(
                AppUrl::getId,
                AppUrl::getUrlValue
            ));
    }

    /*
     * Uno una URL base con una parte final.
     * Lo hago así para evitar errores como barras dobles o falta de barra.
     */
    public String join(String baseUrl, String path) {
        if (baseUrl == null || baseUrl.isBlank()) {
            return path;
        }

        if (path == null || path.isBlank()) {
            return baseUrl;
        }

        String cleanBase = baseUrl.endsWith("/")
            ? baseUrl.substring(0, baseUrl.length() - 1)
            : baseUrl;

        String cleanPath = path.startsWith("/")
            ? path
            : "/" + path;

        return cleanBase + cleanPath;
    }
}

File: LAB/Gramola_Frontend/src/app/app-url.service.ts

/* CHECK
 * Servicio encargado de cargar las URLs de la aplicación desde el backend. 
 * sin esto se apaga
 *
 * Esta clase es importante porque evita que el frontend tenga muchas URLs
 * escritas directamente en el código. Solo dejo una URL inicial para pedir
 * al backend todas las demás URLs guardadas en MySQL.
 *
 * A partir de este servicio, otras clases como UserService, PaymentService
 * o SpotiService pueden pedir la URL que necesitan usando una clave.
 */
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, map, shareReplay } from 'rxjs';

export type AppUrls = Record<string, string>;

@Injectable({
  providedIn: 'root'
})
export class AppUrlService {

  /*
   * Necesito una URL inicial para pedir al backend todas las URLs de MySQL.
   * A partir de aquí, el resto de URLs ya salen de la base de datos.
   */
  private readonly appUrlsEndpoint = 'http://127.0.0.1:8080/appUrls';

  private urls$?: Observable<AppUrls>;

  constructor(private http: HttpClient) {}

  /*
   * Cargo todas las URLs activas desde el backend.
   * Uso shareReplay para no repetir la misma petición cada vez que un servicio necesita una URL.
   */
  getAllUrls(): Observable<AppUrls> {
    if (!this.urls$) {
      this.urls$ = this.http.get<AppUrls>(
        this.appUrlsEndpoint,
        { withCredentials: true }
      ).pipe(
        shareReplay(1)
      );
    }

    return this.urls$;
  }

  /*
   * Devuelvo una URL concreta por su clave.
   * Si falta en base de datos, lanzo error para verlo rápido.
   */
  getUrl(key: string): Observable<string> {
    return this.getAllUrls().pipe(
      map((urls) => {
        const url = urls[key];

        if (!url) {
          throw new Error(`No existe la URL en base de datos: ${key}`);
        }

        return url;
      })
    );
  }
}